src/pybind/mgr/snap_schedule/fs/schedule.py

```diff
@@ -0,0 +1,311 @@
+                               (self.path,))
+                sched_id = c.fetchone()[0]
+                pass
+            db.execute(self.INSERT_SCHEDULE_META,
+                       (sched_id,
+                        self.start.isoformat(),
+                        self.created.isoformat(),
+                        self.repeat,
+                        self.schedule,
+                        self.retention))
+
+    @classmethod
+    def rm_schedule(cls, db, path, repeat, start):
+        with db:
+            cur = db.execute('SELECT id FROM schedules WHERE path = ?',
+                             (path,))
+            row = cur.fetchone()
+
+            if len(row) == 0:
+                log.info(f'no schedule for {path} found')
+                raise ValueError('SnapSchedule for {} not found'.format(path))
+
+            id_ = tuple(row)
+
+            if repeat or start:
+                meta_delete = 'DELETE FROM schedules_meta WHERE schedule_id = ?'
+                delete_param = id_
+                if repeat:
+                    meta_delete += ' AND schedule = ?'
+                    delete_param += (repeat,)
+                if start:
+                    meta_delete += ' AND start = ?'
+                    delete_param += (start,)
+                # maybe only delete meta entry
+                log.debug(f'executing {meta_delete}, {delete_param}')
+                res = db.execute(meta_delete + ';', delete_param).rowcount
+                if res < 1:
+                    raise ValueError(f'No schedule found for {repeat} {start}')
+                db.execute('COMMIT;')
+                # now check if we have schedules in meta left, if not delete
+                # the schedule as well
+                meta_count = db.execute(
+                    'SELECT COUNT() FROM schedules_meta WHERE schedule_id = ?',
+                    id_)
+                if meta_count.fetchone() == (0,):
+                    log.debug(
+                        f'no more schedules left, cleaning up schedules table')
+                    db.execute('DELETE FROM schedules WHERE id = ?;', id_)
+            else:
+                # just delete the schedule CASCADE DELETE takes care of the
+                # rest
+                db.execute('DELETE FROM schedules WHERE id = ?;', id_)
+
+    def report(self):
+        import pprint
+        return pprint.pformat(self.__dict__)
+
+    @property
+    def repeat(self):
+        mult = self.schedule[-1]
+        period = int(self.schedule[0:-1])
+        if mult == 'm':
+            return period * 60
+        elif mult == 'h':
+            return period * 60 * 60
+        elif mult == 'd':
+            return period * 60 * 60 * 24
+        elif mult == 'w':
+            return period * 60 * 60 * 24 * 7
+        else:
+            raise Exception('schedule multiplier not recognized')
+
+    UPDATE_LAST = '''UPDATE schedules_meta
+    SET
+      last = ?,
```

Comment:
this can also be `datetime('now')`

Reply:
I was trying to keep the timestamp creation in the python realm. Motivation here is to have a reliable format and python just seems to handle this better.
Yes we can make sqlite fill in some default dates but this opens up hard to debug room for bugs. Consider the following for example on my laptop (UTC+2):
```
>>> import datetime
>>> datetime.datetime.now().isoformat()
'2020-05-19T15:42:04.905583'
>>> datetime.datetime.now(datetime.timezone.utc).isoformat()
'2020-05-19T13:42:07.115082+00:00'
>>> import sqlite3
>>> con = sqlite3.connect(":memory:")
>>> con.execute('select strftime("%Y-%m-%dT%H:%M.%S","now")').fetchall()[0]
('2020-05-19T13:42.27',)
>>> con.execute('select strftime("%Y-%m-%dT%H:%M.%S","now", "utc")').fetchall()[0]
('2020-05-19T11:42.37',)
```

So I would propose to simply keep timestamp creation in python land, for sanity's sake.